p2p/src/network/pubsub/p2p_network_pubsub_reducer.rs

```diff
@@ -200,29 +217,40 @@ impl P2pNetworkPubsubState {
                     }
                 }
 
-                if let Err(error) = Self::broadcast(dispatcher, global_state) {
-                    bug_condition!(
-                        "Failure when trying to broadcast incoming pubsub message: {error}"
-                    );
+                // This happens if message was already seen
+                let Some(message_content) = message_content else {
+                    return Ok(());
                 };
+                dispatcher.push(P2pNetworkPubsubAction::HandleIncomingMessage {
+                    message,
+                    message_content,
+                    peer_id,
+                });
 
-                if let Some((_, block)) = state.incoming_block.as_ref() {
-                    let best_tip = BlockWithHash::try_new(block.clone())?;
-                    dispatcher.push(P2pPeerAction::BestTipUpdate { peer_id, best_tip });
-                }
-                for (transaction, nonce) in &state.incoming_transactions {
-                    dispatcher.push(P2pChannelsTransactionAction::Libp2pReceived {
-                        peer_id,
-                        transaction: Box::new(transaction.clone()),
-                        nonce: *nonce,
-                    });
-                }
-                for (snark, nonce) in &state.incoming_snarks {
-                    dispatcher.push(P2pChannelsSnarkAction::Libp2pReceived {
-                        peer_id,
-                        snark: Box::new(snark.clone()),
-                        nonce: *nonce,
-                    });
+                Ok(())
+            }
+            P2pNetworkPubsubAction::HandleIncomingMessage {
+                message,
+                message_content,
+                peer_id,
+            } => {
+                let Ok(message_id) =
+                    pubsub_state
+                        .mcache
+                        .put(message, message_content, peer_id, time)
+                else {
+                    bug_condition!("Unable to add message to `mcache`");
+                    return Ok(());
+                };
+
+                let (dispatcher, state) = state_context.into_dispatcher_and_state();
+                let p2p_state: &P2pState = state.substate()?;
+
+                if let Some(callback) = p2p_state.callbacks.on_p2p_pubsub_message_received.clone() {
```

Comment:
can you clarify this? the callback registered there seems to be doing the same thing as the `P2pNetworkPubsubAction::BroadcastValidationCallback` handler? why are both needed? in which case is the `on_p2p_pubsub_message_received` callback not be set?